Fix netllink conntrack handle and add guardrails against zone related entries.

priyankar-jain opened the pull request.

priyankar-jain committed:
During conntrack dump, ignore the entries which does not have IP address set.

There are some cases reported where the conntrack entry received during
the conntrack dump did not had IP address value set. In these cases we
need to ignore these entries.

Signed-off-by: Priyankar Jain <priyankar.jain@nutanix.com>

priyankar-jain committed:
Skip conntrack entries having zone information.

Zone information may change during migration i.e a VM may be in a
different conntrack zone on the destination hypervisor than the source
hypervisor. Hence, we should skip these entries, otherwise it might lead
to an inconsistent state at the destination hypervisor wrt conntrack.

Signed-off-by: Priyankar Jain <priyankar.jain@nutanix.com>

jlevon commented:
what are the situations in which ip addr isn't set? Can the commit message be more explicit?

jlevon commented on src/conntrack.c:
```diff
@@ -108,6 +108,16 @@ conntrack_dump_callback(enum nf_conntrack_msg_type type,
     struct in_addr *src_addr, *dst_addr;
 
     ips_to_migrate = data;
+    if (ct == NULL) {
```
when can this happen

jlevon replied:
as discussed, will remove

priyankar-jain replied:
Done

jlevon commented on src/conntrack.c:
```diff
@@ -108,6 +108,16 @@ conntrack_dump_callback(enum nf_conntrack_msg_type type,
+        return NFCT_CB_CONTINUE;
+    }
+
+    if (nfct_attr_is_set(ct, ATTR_ORIG_IPV4_SRC) <= 0 ||
```
add tests?

priyankar-jain replied:
Done

jlevon commented:
doesn't dropping entries with zone attached cause dataloss (i.e. those entries are not migrated at all) ?

priyankar-jain committed:
Correctly setup netfiler conntrack handle for CT dump.

Currently, the netlink handle used to take conntrack dump is
inadvertently set with subscriptions: NFCT_ALL_CT_GROUPS. This
makes kernel deliver all event messages in the dump handle as well.
This lead to leak of ipv6 based entries in the dump_callback and leading
to segmentation fault.

As per the libnetfilter_conntrack documentation/examples, instead the
subscriptions should be set to 0, so that kernel also provides the dump
related information and honour the family param set during nfct_query.

Signed-off-by: Priyankar Jain <priyankar.jain@nutanix.com>

priyankar-jain committed:
Updated README

Signed-off-by: Priyankar Jain <priyankar.jain@nutanix.com>

jlevon commented on README.md:
```diff
@@ -83,6 +83,13 @@ For eg:
+Following features are not yet supported:
+ - Conntrack entries having IPv6 information will not be migrated.
+ - Conntrack entries having zone information will not be migrated
+   since there is not guarntee that on the destination the same zone
```
no guarantee

priyankar-jain replied:
Done

jlevon commented on src/main.c:
```diff
@@ -136,7 +136,7 @@ dump_conntrack(GHashTable *ips_to_migrate)
 {
     struct nfct_handle *handle;
 
```
add a comment describing what we discussed?

priyankar-jain replied:
Done

priyankar-jain committed:
Addressed review comments and added UTs.

Signed-off-by: Priyankar Jain <priyankar.jain@nutanix.com>

jlevon commented on README.md:
```diff
@@ -83,6 +83,13 @@ For eg:
+Following features are not yet supported:
+ - Conntrack entries having IPv6 information will not be migrated.
+ - Conntrack entries having zone information will not be migrated
+   since there is not guarantee that on the destination the same zone
```
still needs fixing

priyankar-jain replied:
Done

jlevon commented on src/main.c:
```diff
@@ -136,7 +136,9 @@ dump_conntrack(GHashTable *ips_to_migrate)
 
-    handle = nfct_open(CONNTRACK, NFCT_ALL_CT_GROUPS);
+    // Set the subscriptions for this handle to 0 since we are going to
+    // explicily request dump on this socket.
```
explicitly

priyankar-jain replied:
Done

jlevon reviewed (commented):
just some typos.

Please make sure to squash the commits into one logical change with a fixed commit message, thanks.

priyankar-jain committed:
Fixed some typos.

Signed-off-by: Priyankar Jain <priyankar.jain@nutanix.com>

jlevon commented on src/conntrack.c:
```diff
@@ -108,11 +108,32 @@ conntrack_dump_callback(enum nf_conntrack_msg_type type,
+    if (nfct_attr_is_set(ct, ATTR_ORIG_IPV4_SRC) <= 0 ||
+        nfct_attr_is_set(ct, ATTR_ORIG_IPV4_DST) <= 0) {
+        char *buf = g_malloc0(1024);
+        if (buf != NULL) {
```
g_malloc0 can't fail like this

jlevon replied:
please, also, add a little helper function

priyankar-jain replied:
Done.

jlevon commented on src/conntrack.c:
```diff
@@ -230,6 +251,28 @@ conntrack_events_callback(const struct nlmsghdr *nlh, void *data)
+
+    if (nfct_attr_is_set(ct, ATTR_ORIG_IPV4_SRC) <= 0 ||
+        nfct_attr_is_set(ct, ATTR_ORIG_IPV4_DST) <= 0) {
+        char *buf = g_malloc0(1024);
```
ditto

jlevon commented on src/conntrack.c:
```diff
@@ -552,6 +595,25 @@ delete_conntrack_dump_callback(enum nf_conntrack_msg_type type,
 
+    if (nfct_attr_is_set(ct, ATTR_ORIG_IPV4_SRC) <= 0 ||
+        nfct_attr_is_set(ct, ATTR_ORIG_IPV4_DST) <= 0) {
+        char *buf = g_malloc0(1024);
```
ditto

priyankar-jain committed:
Added a helper fn for validating CT entry to reduce code duplication.

Signed-off-by: Priyankar Jain <priyankar.jain@nutanix.com>

jlevon reviewed (approved):
thanks, please remember to squash
